### ISSUE TITLE
getTargetRanges of `beforeinput` on EditContext editing host

### ISSUE THREAD

**Issue by masayuki-nakano** [w3c/edit-context#73]
Should it always return empty array? In my understanding, browsers do not manage selection range as deleting range in EditContext editing host. Therefore, I think that the range should be considered by the web apps themselves.

On the other hand, [this example](https://w3c.github.io/edit-context/#example-how-beforeinput-can-be-used-to-catch-insertreplacementtext-to-apply-spelling-changes) uses the result of `getTargetRanges`...

**Comment by dandclark** [comment]
As pointed out by @johanneswilm and as stated in https://w3c.github.io/edit-context/#edit-context-differences, the user agent will still handle selection and caret navigation in an EditContext. So I think it makes sense that `getTargetRanges` still works the normal way in EditContext. This is necessary for authors to be able to handle things like replacement from spellcheckers, which would trigger `beforeinput` with type `"insertReplacementText"`. The author will need the result of `getTargetRanges()` in that case.

**Comment by masayuki-nakano** [comment]
My main point is about the cases deleting text from collapsed range. It requires a lot of computation to consider the range to delete even though web apps finally consider the behavior and the result is not consistent between browsers. See [current test result](https://wpt.fyi/results/input-events?label=master&label=experimental&aligned). The amount how much the user wants to delete can be detected is notified by [.inputType](https://w3c.github.io/input-events/#interface-InputEvent-Attributes).

In other words, if `getTargetRanges` is really required for any `inputType` values, the detail of `getTargetRanges` should be defined first, but it could be impossible because there are too many situations.

**Comment by johanneswilm** [comment]
@masayuki-nakano Can you not wait with calculating the target ranges until the `getTargetRanges()` method is called? Web apps that want to decide the target ranges themselves and don't require any input from the browser then don't have to call the method and so the calculation never takes place.

The exact ranges that a browser-provided target range covers will be OS/browser dependent as is also now the case. That's why we didn't specify it any further in the case of contenteditable, and I'm imaging in this case you can just run the same code that you are running for contenteditable to calculate them. Is that not the case?

**Comment by masayuki-nakano** [comment]
> @masayuki-nakano Can you not wait with calculating the target ranges until the `getTargetRanges()` method is called? Web apps that want to decide the target ranges themselves and don't require any input from the browser then don't have to call the method and so the calculation never takes place.

Good idea for the `beforeinput` event performance (even for `contenteditable`). However, one of the problem is, if `EditContext` works without `getTargetRanges`, we can avoid to use the builtin HTML editor, but it becomes impossible to port the computing target range command. And there is another problem. The spec needs to define **when** the result should be computed. The DOM tree can be modified even during the `beforeinput` propagation or even after the propagation. In the latter case, Chrome behavior is broken, IIRC.

> The exact ranges that a browser-provided target range covers will be OS/browser dependent as is also now the case. That's why we didn't specify it any further in the case of contenteditable, and I'm imaging in this case you can just run the same code that you are running for contenteditable to calculate them. Is that not the case?

I think that it does not make sense for web apps to use `getTargetRanges` without ignoring a lot of odd result because [the result](https://wpt.fyi/results/input-events?q=get-target-ranges&run_id=5088436001701888&run_id=4800231314554880&run_id=5072135761952768&run_id=5067839888687104) means they are not just incompatible especially when deleting parent blocks, etc. E.g., Type `Backspace` in `<ul><li>{}<br></li></ul>`. Chrome and Firefox delete the `<ul>` entirely, but Chrome returns the collapsed range. I agree with `getTargetRanges` is useful for text deletion, but current spec does not define the behavior for most cases and browsers implement it too roughly. (Especially, when I implement `beforeinput` as the last implementer, there were no meaningful `getTargetRange` tests in WPT...)

Even if browsers should return useful ranges for text deletion, I think that the spec should define the behavior clearly. E.g., word amount backward deletion at `<div><b>ab</b><i>cd[]</i></div>`, staring from `{div,0}`, `{b,0}` or `{#text,0}`?, and ending at `{div,2}`, `{i,1}` or `#text,2`? So, I think that supporting useful `getTargetRanges` across browsers require a lot of work and it seems that it may not be reasonable for actual usage.

**Comment by johanneswilm** [comment]
> Even if browsers should return useful ranges for text deletion, I think that the spec should define the behavior clearly.

We can try to discuss this point again. Last time we concluded that it's just not possible as different platforms have different ideas about what should be the right target range. This seems like a general question related to `getTargetRanges()` also in contenteditable and not just related to `EditContext`. Do I understand that correctly? If yes, should we move this issue to the input events spec?

I think that the target ranges are more important for some `inputType`s than others. When using deletion forward/backward, most JavaScript apps should be able to figure out what they want to do without using target ranges by looking at the `inputType` and  the current selection. On the other hand, in the case of using the built-in spell checker, it's important to know which range is to be replaced with the incoming text and because it doesn't replace the current selection, there really is no other way to know where to insert the text without looking at target ranges.

**Comment by johanneswilm** [comment]
> I think it will not end up mattering much because most developers would not look at `getTargetRanges()` for deletion commands since they would compute the extent of the deletion themselves.

Actually, we received reports from JavaScript editors previously that they need this information as there can be platform dependent differences for things like deleting a word or line forward or backward that would require re-implementing a lot in JavaScript that is already in the browser.

Looking at the way target ranges are used in an editor like CKEditor5 that is working mainly using the `beforeinput` event, it looks like they are not always that helpful though [1].

As this seems to be a general issue with target ranges, I will move this issue to the input events spec. 

[1] https://github.com/ckeditor/ckeditor5/blob/331d1e7a04729284efbb55937fd97a452488dc8d/packages/ckeditor5-engine/src/view/observer/inputobserver.ts#L82-L83
